File: tesis-slm-quechua/src/inferencia/enrutador_lenguaje.py
from __future__ import annotations

import re

QU_MARKERS = {
    "imaynata",
    "ima",
    "mayqin",
    "allin",
    "yachay",
    "yachachiq",
    "wawa",
    "runa",
    "pacha",
    "yaku",
    "inti",
    "waspi",
    "kay",
    "chay",
    "ñuqa",
    "nuqa",
    "qam",
    "pay",
    "noqanchis",
    "rimay",
    "qillqay",
    "t'ikray",
    "kawsay",
    "pukllay",
}


def detect_language(text: str) -> str:
    tokens = re.findall(r"[A-Za-zÁÉÍÓÚÜÑáéíóúüñ']+", text.lower())
    if not tokens:
        return "unknown"

    qu_hits = sum(1 for t in tokens if t in QU_MARKERS or t.endswith(("chu", "paq", "kuna", "ykuna")))
    es_hits = sum(1 for t in tokens if t in {"que", "es", "la", "el", "de", "para", "como", "porque", "una", "los"})

    if qu_hits >= 2 and es_hits >= 2:
        return "mixed_es_qu"
    if qu_hits >= 2 and qu_hits > es_hits:
        return "qu_collao"
    if es_hits >= 1 or qu_hits == 0:
        return "es"
    if qu_hits >= 1:
        return "qu_collao"
    return "unknown"


detectar_lenguaje = detect_language

File: tesis-slm-quechua/apps/pwa/app/rag/enrutadorLenguaje.ts
// Puerto literal de src/inferencia/enrutador_lenguaje.py

export type IdiomaDetectado = 'es' | 'qu_collao' | 'mixed_es_qu' | 'unknown';

const QU_MARKERS = new Set([
  'imaynata',
  'ima',
  'mayqin',
  'allin',
  'yachay',
  'yachachiq',
  'wawa',
  'runa',
  'pacha',
  'yaku',
  'inti',
  'waspi',
  'kay',
  'chay',
  'ñuqa',
  'nuqa',
  'qam',
  'pay',
  'noqanchis',
  'rimay',
  'qillqay',
  "t'ikray",
  'kawsay',
  'pukllay',
]);

const ES_MARKERS = new Set([
  'que',
  'es',
  'la',
  'el',
  'de',
  'para',
  'como',
  'porque',
  'una',
  'los',
]);

const QU_SUFFIXES = ['chu', 'paq', 'kuna', 'ykuna'];

export function detectLanguage(text: string): IdiomaDetectado {
  const tokens = text.toLowerCase().match(/[A-Za-zÁÉÍÓÚÜÑáéíóúüñ']+/g);
  if (!tokens || tokens.length === 0) {
    return 'unknown';
  }

  let quHits = 0;
  let esHits = 0;
  for (const t of tokens) {
    if (QU_MARKERS.has(t) || QU_SUFFIXES.some((suf) => t.endsWith(suf))) {
      quHits += 1;
    }
    if (ES_MARKERS.has(t)) {
      esHits += 1;
    }
  }

  if (quHits >= 2 && esHits >= 2) {
    return 'mixed_es_qu';
  }
  if (quHits >= 2 && quHits > esHits) {
    return 'qu_collao';
  }
  if (esHits >= 1 || quHits === 0) {
    return 'es';
  }
  if (quHits >= 1) {
    return 'qu_collao';
  }
  return 'unknown';
}
